### ISSUE TITLE
Error during package installation ('maximum recursion depth exceeded',)

### ISSUE THREAD

**Issue by fdasilva59** [conda/conda#6994]
Hi,

**I'm submitting a...**
  - [x] bug report
  - [ ] feature request

## Current Behavior

On a fresh conda installation, when creating a conda environment with "large" packages like mkl-2018.0.1,  
which is a dependancy requirement for Numpy, the download of some packages stops quickly, and conda fails with "maximum recursion depth exceeded" errors.

Notes:
- I am behind an entreprise proxy (for which I don't have much info). But I was using conda since one year without any troubles on that same environment. I discovered the problem around Week 8. I was able to reproduce the problem on other servers behind the same proxy, or on a different proxy (for which I don't details neither) 
- there is no issue to donwload stuff from internet from that server. I can manually download the faulty packages from https://repo.continuum.io/pkgs/main/linux-64/ repository with curl or wget whithout any troubles. But that fails with conda for certain packages
- this seems to concern large packages. The problem now occurs systematically with mkl-2018.0.1
- If the packages to install are already in the local cache, the problem might not be visible
- Currently I am testing with latest version of conda package and latest  version of Miniconda3 installer. 
But I have tested also with older versions that proved to be working before for me. Same error.

I wonder if this could be similar/related to https://github.com/conda/conda/issues/5032 ?

### Steps to Reproduce

```
# In a CentOS7 docker image or VM
curl -o /tmp/Miniconda3-latest-Linux-x86_64.sh https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh 
chmod u+x /tmp/Miniconda3-latest-Linux-x86_64.sh
/bin/sh /tmp/Miniconda3-latest-Linux-x86_64.sh -f -b -p /usr/local/
conda update -y conda
conda create -y --prefix /opt/DataScience python=3.6 numpy

```

Here are the logs with no debug
```
[root@4bd8550ee6e4 / ]$ conda create -y --prefix /opt/DataScience python=3.6 numpy
Solving environment: done

## Package Plan ##

  environment location: /opt/DataScience

  added / updated specs: 
    - numpy
    - python=3.6


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    intel-openmp-2018.0.0      |       hc7b2577_8         618 KB
    pip-9.0.1                  |           py36_5         2.2 MB
    setuptools-38.5.1          |           py36_0         525 KB
    numpy-1.14.1               |   py36h3dfced4_1         4.0 MB
    mkl-2018.0.1               |       h19d6760_4       184.7 MB
    libgfortran-ng-7.2.0       |       h9f7466a_2         349 KB
    ------------------------------------------------------------
                                           Total:       192.3 MB

The following NEW packages will be INSTALLED:

    ca-certificates: 2017.08.26-h1d4fec5_0
    certifi:         2018.1.18-py36_0     
    intel-openmp:    2018.0.0-hc7b2577_8  
    libedit:         3.1-heed3624_0       
    libffi:          3.2.1-hd88cf55_4     
    libgcc-ng:       7.2.0-h7cc24e2_2     
    libgfortran-ng:  7.2.0-h9f7466a_2     
    libstdcxx-ng:    7.2.0-h7a57d05_2     
    mkl:             2018.0.1-h19d6760_4  
    ncurses:         6.0-h9df7e31_2       
    numpy:           1.14.1-py36h3dfced4_1
    openssl:         1.0.2n-hb7f436b_0    
    pip:             9.0.1-py36_5         
    python:          3.6.4-hc3d631a_1     
    readline:        7.0-ha6073c6_4       
    setuptools:      38.5.1-py36_0        
    sqlite:          3.22.0-h1bed415_0    
    tk:              8.6.7-hc745277_3     
    wheel:           0.30.0-py36hfd4bba0_1
    xz:              5.2.3-h55aa19d_2     
    zlib:            1.2.11-ha838bed_2    


Downloading and Extracting Packages
intel-openmp 2018.0.0: ################################################################################################################################################################################## | 100% 
pip 9.0.1: ############################################################################################################################################################################################## | 100% 
setuptools 38.5.1: ###################################################################################################################################################################################### | 100% 
numpy 1.14.1: ########################################################################################################################################5                                                   |  73% 
mkl 2018.0.1: #1                                                                                                                                                                                          |   1% 
libgfortran-ng 7.2.0: ################################################################################################################################################################################### | 100% 

RecursionError('maximum recursion depth exceeded',)
RecursionError('maximum recursion depth exceeded',)

[root@4bd8550ee6e4 / ]$ ls -l  --block-size=M /usr/local/pkgs/mkl-2018.0.1-h19d6760_4.tar.bz2 
-rw-r--r--. 1 root root 2M Mar  6 13:43 /usr/local/pkgs/mkl-2018.0.1-h19d6760_4.tar.bz2

[root@4bd8550ee6e4 / ]$ curl -o /dev/null https://repo.continuum.io/pkgs/main/linux-64/mkl-2018.0.1-h19d6760_4.tar.bz2
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  184M  100  184M    0     0  1393k      0  0:02:15  0:02:15 --:--:-- 1066k

```

Here are selected logs running the same command but with the `-v -v -v --debug` flags :

``` 
[...]

DEBUG urllib3.connectionpool:_make_request(396): https://repo.continuum.io:443 "GET /pkgs/main/linux-64/libgfortran-ng-7.2.0-h9f7466a_2.tar.bz2 HTTP/1.1" 200 357812
DEBUG conda.gateways.connection.download:download(45): 
>>GET /pkgs/main/linux-64/libgfortran-ng-7.2.0-h9f7466a_2.tar.bz2 HTTPS
> User-Agent: conda/4.4.11 requests/2.18.4 CPython/3.6.4 Linux/3.10.0-514.26.2.el7.x86_64 centos/7 glibc/2.17
> Accept: */*
> Accept-Encoding: gzip, deflate
> Connection: keep-alive

<<HTTPS 200 OK
< Cache-Control: public, max-age=30
< CF-Cache-Status: HIT
< CF-RAY: 3f751c9d2f44a845-CDG
< Content-Type: application/x-tar
< Date: Tue, 06 Mar 2018 13:19:03 GMT
< ETag: "f5ae707dd1843643224d5946e901a960"
< Expect-CT: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
< Expires: Tue, 06 Mar 2018 13:19:33 GMT
< Last-Modified: Fri, 06 Oct 2017 20:51:24 GMT
< Server: cloudflare
< Set-Cookie: __cfduid=dae4a8f2adca3cee1745f0aa2bd6095231520342343; expires=Wed, 06-Mar-19 13:19:03 GMT; path=/; domain=.continuum.io; HttpOnly
< Vary: Accept-Encoding
< x-amz-id-2: TILwoDuWMTpQMmpsWuta8Ke9QdjYC9rBCZFl4iaH6uZE1xgMP7Z9VW/iBBo4haPQHILY+Lqv0ig=
< x-amz-request-id: BB821C93E18308EB
< x-amz-version-id: 4DEXQP.o3kUnG3BjBZbSjexbApZioxXi
< Content-Length: 357812
< Connection: keep-alive
< Elapsed: 00:00.103545

DEBUG conda.gateways.disk.create:extract_tarball(132): extracting /usr/local/pkgs/libgfortran-ng-7.2.0-h9f7466a_2.tar.bz2
  to /usr/local/pkgs/libgfortran-ng-7.2.0-h9f7466a_2

[...]

DEBUG urllib3.connectionpool:_make_request(396): https://repo.continuum.io:443 "GET /pkgs/main/linux-64/numpy-1.14.1-py36h3dfced4_1.tar.bz2 HTTP/1.1" 200 4239251
DEBUG conda.gateways.connection.download:download(45): 
>>GET /pkgs/main/linux-64/numpy-1.14.1-py36h3dfced4_1.tar.bz2 HTTPS
> User-Agent: conda/4.4.11 requests/2.18.4 CPython/3.6.4 Linux/3.10.0-514.26.2.el7.x86_64 centos/7 glibc/2.17
> Accept: */*
> Accept-Encoding: gzip, deflate
> Connection: keep-alive
> Cookie: __cfduid=dae4a8f2adca3cee1745f0aa2bd6095231520342343

<<HTTPS 200 OK
< Cache-Control: public, max-age=30
< CF-Cache-Status: HIT
< CF-RAY: 3f751cb2881fa845-CDG
< Content-Type: application/x-tar
< Date: Tue, 06 Mar 2018 13:19:06 GMT
< ETag: "007c2dd69dc05437ba602ed86e3ddc98"
< Expect-CT: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
< Expires: Tue, 06 Mar 2018 13:19:36 GMT
< Last-Modified: Wed, 21 Feb 2018 15:52:35 GMT
< Server: cloudflare
< Vary: Accept-Encoding
< x-amz-id-2: p0YW03SyhGitGloprmJ0ua8MfgkGn7SvPTfahG8ZcUahesmRxYCOyjx4xWm4ZKbuDyesvdlaLFw=
< x-amz-request-id: 15D728DA92F109E9
< x-amz-version-id: GvsaBjIdiF4IxQ4rFy4Dg63y2EQaHvVy
< Content-Length: 4239251
< Connection: keep-alive
< Elapsed: 00:00.019594

DEBUG conda.gateways.disk.create:extract_tarball(132): extracting /usr/local/pkgs/numpy-1.14.1-py36h3dfced4_1.tar.bz2
  to /usr/local/pkgs/numpy-1.14.1-py36h3dfced4_1
DEBUG urllib3.connectionpool:_make_request(396): https://repo.continuum.io:443 "GET /pkgs/main/linux-64/intel-openmp-2018.0.0-hc7b2577_8.tar.bz2 HTTP/1.1" 200 632408
DEBUG conda.gateways.connection.download:download(45): 
>>GET /pkgs/main/linux-64/intel-openmp-2018.0.0-hc7b2577_8.tar.bz2 HTTPS
> User-Agent: conda/4.4.11 requests/2.18.4 CPython/3.6.4 Linux/3.10.0-514.26.2.el7.x86_64 centos/7 glibc/2.17
> Accept: */*
> Accept-Encoding: gzip, deflate
> Connection: keep-alive
> Cookie: __cfduid=dae4a8f2adca3cee1745f0aa2bd6095231520342343

<<HTTPS 200 OK
< Cache-Control: public, max-age=30
< CF-Cache-Status: HIT
< CF-RAY: 3f751cd40fcea845-CDG
< Content-Type: application/x-tar
< Date: Tue, 06 Mar 2018 13:19:12 GMT
< ETag: "658ff659cd4178ae7b41a613188b025b"
< Expect-CT: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
< Expires: Tue, 06 Mar 2018 13:19:42 GMT
< Last-Modified: Wed, 22 Nov 2017 19:08:14 GMT
< Server: cloudflare
< Vary: Accept-Encoding
< x-amz-id-2: WKLu+k1nHVuYW7A3YGkHl19tLwi4NKXYBI/apQeezeyc05cpt/TP0duYDpwhDgPFZHtsNQShtfk=
< x-amz-request-id: 825E37D746D8FCD7
< x-amz-version-id: ZBFQdOnCm0Bhaz7GDuI9mDLXuGPOh_td
< Content-Length: 632408
< Connection: keep-alive
< Elapsed: 00:00.040946

DEBUG conda.gateways.disk.create:extract_tarball(132): extracting /usr/local/pkgs/intel-openmp-2018.0.0-hc7b2577_8.tar.bz2
  to /usr/local/pkgs/intel-openmp-2018.0.0-hc7b2577_8
DEBUG urllib3.connectionpool:_make_request(396): https://repo.continuum.io:443 "GET /pkgs/main/linux-64/mkl-2018.0.1-h19d6760_4.tar.bz2 HTTP/1.1" 200 193625096
DEBUG conda.gateways.connection.download:download(45): 
>>GET /pkgs/main/linux-64/mkl-2018.0.1-h19d6760_4.tar.bz2 HTTPS
> User-Agent: conda/4.4.11 requests/2.18.4 CPython/3.6.4 Linux/3.10.0-514.26.2.el7.x86_64 centos/7 glibc/2.17
> Accept: */*
> Accept-Encoding: gzip, deflate
> Connection: keep-alive
> Cookie: __cfduid=dae4a8f2adca3cee1745f0aa2bd6095231520342343

<<HTTPS 200 OK
< Cache-Control: public, max-age=30
< CF-Cache-Status: HIT
< CF-RAY: 3f751cdfbb73a845-CDG
< Content-Type: application/x-tar
< Date: Tue, 06 Mar 2018 13:19:13 GMT
< ETag: "c79cdd612353b84d9138311b1e727480-24"
< Expect-CT: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
< Expires: Tue, 06 Mar 2018 13:19:43 GMT
< Last-Modified: Wed, 22 Nov 2017 19:08:14 GMT
< Server: cloudflare
< Vary: Accept-Encoding
< x-amz-id-2: IXKgNUywyozLPy3ibeeHZ9hbaeC6eyJXnQVsxgL56UYUwT4Dsa2icA5HIQf/CsZ58dcTyfyS/MA=
< x-amz-request-id: A24679D22B61EF94
< x-amz-version-id: D4bcGCyZkJYFpzNGKD35vtaPKAjD33bR
< Content-Length: 193625096
< Connection: keep-alive
< Elapsed: 00:00.017244

DEBUG conda.core.package_cache:execute(586): RecursionError('maximum recursion depth exceeded',)
NoneType: None
DEBUG conda.common.signals:signal_handler(56): de-registering handler for Signals.SIGABRT
DEBUG conda.common.signals:signal_handler(56): de-registering handler for Signals.SIGINT
DEBUG conda.common.signals:signal_handler(56): de-registering handler for Signals.SIGTERM
DEBUG conda.common.signals:signal_handler(56): de-registering handler for Signals.SIGQUIT
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/conda/exceptions.py", line 789, in __call__
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/conda/cli/main.py", line 78, in _main
    exit_code = do_call(args, p)
  File "/usr/local/lib/python3.6/site-packages/conda/cli/conda_argparse.py", line 77, in do_call
    exit_code = getattr(module, func_name)(args, parser)
  File "/usr/local/lib/python3.6/site-packages/conda/cli/main_create.py", line 11, in execute
    install(args, parser, 'create')
  File "/usr/local/lib/python3.6/site-packages/conda/cli/install.py", line 255, in install
    handle_txn(progressive_fetch_extract, unlink_link_transaction, prefix, args, newenv)
  File "/usr/local/lib/python3.6/site-packages/conda/cli/install.py", line 281, in handle_txn
    progressive_fetch_extract.execute()
  File "/usr/local/lib/python3.6/site-packages/conda/core/package_cache.py", line 590, in execute
    raise CondaMultiError(exceptions)
conda.CondaMultiError: maximum recursion depth exceeded

```

## Expected Behavior

The packages download should not fails with conda (troncated bzip2 archive in local cache)

## Environment Information
<details open><summary><code>`conda info`</code></summary><p>

```
active environment : None
       user config file : /root/.condarc
 populated config files : 
          conda version : 4.4.11
    conda-build version : not installed
         python version : 3.6.4.final.0
       base environment : /usr/local  (writable)
           channel URLs : https://repo.continuum.io/pkgs/main/linux-64
                          https://repo.continuum.io/pkgs/main/noarch
                          https://repo.continuum.io/pkgs/free/linux-64
                          https://repo.continuum.io/pkgs/free/noarch
                          https://repo.continuum.io/pkgs/r/linux-64
                          https://repo.continuum.io/pkgs/r/noarch
                          https://repo.continuum.io/pkgs/pro/linux-64
                          https://repo.continuum.io/pkgs/pro/noarch
          package cache : /usr/local/pkgs
                          /root/.conda/pkgs
       envs directories : /usr/local/envs
                          /root/.conda/envs
               platform : linux-64
             user-agent : conda/4.4.11 requests/2.18.4 CPython/3.6.4 Linux/3.10.0-514.26.2.el7.x86_64 centos/7 glibc/2.17
                UID:GID : 0:0
             netrc file : None
           offline mode : False
```
</p></details>

<details open><summary><code>`conda config --show-sources`</code></summary><p>

```
[root@4bd8550ee6e4 / ]$ conda config --show-sources

[root@4bd8550ee6e4 / ]$
```
</p></details>

<details><summary><code>`conda list --show-channel-urls`</code></summary><p>

```
[root@4bd8550ee6e4 / ]$ conda list --show-channel-urls
# packages in environment at /usr/local:
#
# Name                    Version                   Build  Channel
asn1crypto                0.24.0                   py36_0    defaults
ca-certificates           2017.08.26           h1d4fec5_0    defaults
certifi                   2018.1.18                py36_0    defaults
cffi                      1.11.4           py36h9745a5d_0    defaults
chardet                   3.0.4            py36h0f667ec_1    defaults
conda                     4.4.11                   py36_0    defaults
conda-env                 2.6.0                h36134e3_1    defaults
cryptography              2.1.4            py36hd09be54_0    defaults
idna                      2.6              py36h82fb2a8_1    defaults
libedit                   3.1                  heed3624_0    defaults
libffi                    3.2.1                hd88cf55_4    defaults
libgcc-ng                 7.2.0                h7cc24e2_2    defaults
libstdcxx-ng              7.2.0                h7a57d05_2    defaults
ncurses                   6.0                  h9df7e31_2    defaults
openssl                   1.0.2n               hb7f436b_0    defaults
pip                       9.0.1            py36h6c6f9ce_4    defaults
pycosat                   0.6.3            py36h0a5515d_0    defaults
pycparser                 2.18             py36hf9f622e_1    defaults
pyopenssl                 17.5.0           py36h20ba746_0    defaults
pysocks                   1.6.7            py36hd97a5b1_1    defaults
python                    3.6.4                hc3d631a_1    defaults
readline                  7.0                  ha6073c6_4    defaults
requests                  2.18.4           py36he2e5f8d_1    defaults
ruamel_yaml               0.15.35          py36h14c3975_1    defaults
setuptools                38.4.0                   py36_0    defaults
six                       1.11.0           py36h372c433_1    defaults
sqlite                    3.22.0               h1bed415_0    defaults
tk                        8.6.7                hc745277_3    defaults
urllib3                   1.22             py36hbe7ace6_0    defaults
wheel                     0.30.0           py36hfd4bba0_1    defaults
xz                        5.2.3                h55aa19d_2    defaults
yaml                      0.1.7                had09818_2    defaults
zlib                      1.2.11               ha838bed_2    defaults

```
</p></details>

**Comment by mbargull** [comment]
Thanks for the detailed report! Looks like a duplicate of #6960.

**Comment by fdasilva59** [comment]
In order to avoid being stuck at installing/using python virtual environments, I tried to use pipenv, but it's nowhere even close to conda (well maybe except for the the Pipfile requirement file which is automatically upded, which is a cool feature). So I have tried to find a conda workaround. 

As stated in the debug logs, the problem occurs in the download function of [download.py](https://github.com/conda/conda/blob/master/conda/gateways/connection/download.py)

This function appears to download packages in streamed bytes chunks, checks MD5, ... and for some reasons it now fails for me (at home (no proxy) or at work (behind an enterprise proxy). I did not have time to debug further the issue, but I came across this simple hack to hijack the download function and force the download  of the conda packages files \:

Should anyone being interested, if you want to try this workaround, just insert this code at the very begining of the download function \:

```
@time_recorder("download")
def download(url, target_full_path, md5sum, progress_update_callback=None):
    
    # Dirty patch/hack to hijack and fix conda package download
    print("\nHACK: url=",url, " target_full_path", target_full_path)
    from urllib.request import urlretrieve
    urlretrieve(url, target_full_path)
    return

    [...] (Original source code)
```

**Comment by CSchulzeTLK** [comment]
I'm having the same problem with windows.

To reproduce under windows you can call:
```
set home=b:\
conda install mypackage
```

Changing the home environment changes the output of expanduser("~") in conda\base\context.py

I debugged the problem and found that the dirname() of "b:\\" is "b:\\". If you cannot access it due to missing rights etc. then the isdir is False. The result is an infinite recursion.

**Comment by CSchulzeTLK** [comment]
Maybe an error message?

```
def mkdir_p_sudo_safe(path):
    if isdir(path):
        return
    base_dir = dirname(path)
    if not isdir(base_dir):
        ################ ->
        if (base_dir == path):
            log.error("cannot access "+path)
            raise
        ################ <-
        mkdir_p_sudo_safe(base_dir)
    log.trace('making directory %s', path)
    mkdir(path)
    if not on_win and os.environ.get('SUDO_UID') is not None:
        uid = int(os.environ['SUDO_UID'])
        gid = int(os.environ.get('SUDO_GID', -1))
        log.trace("chowning %s:%s %s", uid, gid, path)
        os.chown(path, uid, gid)
```

->

Preparing transaction: done
Verifying transaction: - ERROR conda.gateways.disk:mkdir_p_sudo_safe(75): cannot access b:\

**Comment by mbargull** [comment]
@fdasilva59, thanks for tracking this down, should be helpful information when looking to fix this.
@CSchulzeTLK, you are encountering a different issue, see https://github.com/conda/conda/issues/6937#issuecomment-368952299.

**Comment by kalefranz** [comment]
Closing this issue as a duplicate.  There are two issues here.  The first is a duplicate of #6960.  The second should be addressed with resolution of #6937.

**Comment by QuantumPhy** [comment]
in case it may help someone 
it's work for me , i add this 
```
import sys 
sys.setrecursionlimit(limit)
```
to "download.py" located in [...]conda/gateways/connection/
